### PR TITLE
Xcode frameworks

### PR DESCRIPTION
**Pull request by kalmiya**
Initial version of supporting xcodeproj frameworks
- Adds the frameworks as a "Link Binary with Libraries" build-phase.
- Frameworks also added as a group to the project (just like sources are)
- no need to specify custom linker flags anymore for xcodeproj 

example meson.build file :

```meson
dep_main += [dependency('appleframeworks',
  modules : ['Foundation', 'AppKit', 'IOKIT', 'QuartzCore', 'OpenGL', 'GLUT', 'OpenAL'],
  required : true)]
```

previously needed to sth like this for the xcode-backend:
```meson
 	link_extra_args = '-Wl, ' 
 	link_extra_args += '-lfreetype -L /opt/X11/lib '
 	link_extra_args += '-framework Foundation ' 
 	link_extra_args += '-framework AppKit '
 	link_extra_args += '-framework IOKIT '
        ...
```

see also: https://groups.google.com/forum/#!topic/mesonbuild/CE-GOuh2NvQ

### CODE REVIEW

File: mesonbuild/backend/xcodebackend.py [hunk 1]
```diff
@@ -133,7 +136,14 @@ def generate_buildstylemap(self):
         self.buildstylemap = {'debug': self.gen_id()}
 
     def generate_build_phase_map(self):
-        self.buildphasemap = {}
+        # Probably should be a vector of maps, one vector-entry for each native_targets (e.g. macos/ios/...)
```
Comment (ignatenkobrain) [inline mesonbuild/backend/xcodebackend.py]: __[flake8]__
* [W291] trailing whitespace

File: mesonbuild/backend/xcodebackend.py [hunk 2]
```diff
@@ -133,7 +136,14 @@ def generate_buildstylemap(self):
         self.buildstylemap = {'debug': self.gen_id()}
 
     def generate_build_phase_map(self):
-        self.buildphasemap = {}
+        # Probably should be a vector of maps, one vector-entry for each native_targets (e.g. macos/ios/...) 
+        # and each native-target can have their own Frameworks-section (with its own unique-id)
+        self.buildphasemap = {}
```
Comment (ignatenkobrain) [inline mesonbuild/backend/xcodebackend.py]: __[flake8]__
* [W291] trailing whitespace

File: mesonbuild/backend/xcodebackend.py [hunk 3]
```diff
@@ -222,7 +242,14 @@ def generate_pbx_build_file(self):
         self.ofile.write('\n/* Begin PBXBuildFile section */\n')
         templ = '%s /* %s */ = { isa = PBXBuildFile; fileRef = %s /* %s */; settings = { COMPILER_FLAGS = "%s"; }; };\n'
         otempl = '%s /* %s */ = { isa = PBXBuildFile; fileRef = %s /* %s */;};\n'
+
         for t in self.build.targets.values():
+
+            for dep in t.get_external_deps():
+                if isinstance(dep, dependencies.AppleFrameworks):
+                    for f in dep.frameworks:
+                        self.ofile.write('%s /* %s.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = %s /* %s.framework */; };\n' %(self.native_frameworks[f], f, self.native_frameworks_fileref[f], f))
```
Comment (ignatenkobrain) [inline mesonbuild/backend/xcodebackend.py]: __[flake8]__
* [E225] missing whitespace around operator

File: mesonbuild/backend/xcodebackend.py [hunk 4]
```diff
@@ -276,6 +303,13 @@ def generate_pbx_container_item_proxy(self):
 
     def generate_pbx_file_reference(self):
         self.ofile.write('\n/* Begin PBXFileReference section */\n')
+
+        for t in self.build.targets.values():
+            for dep in t.get_external_deps():
+                if isinstance(dep, dependencies.AppleFrameworks):
+                    for f in dep.frameworks:
+                        self.ofile.write('%s /* %s.framework */ = {isa = PBXFileReference; lastKnownFileType = wrapper.framework; name = %s.framework; path = System/Library/Frameworks/%s.framework; sourceTree = SDKROOT; };\n' %(self.native_frameworks_fileref[f], f, f, f))
```
Comment (ignatenkobrain) [inline mesonbuild/backend/xcodebackend.py]: __[flake8]__
* [E225] missing whitespace around operator

File: mesonbuild/backend/xcodebackend.py [hunk 5]
```diff
@@ -276,6 +303,13 @@ def generate_pbx_container_item_proxy(self):
...
+
```
Comment (ignatenkobrain) [inline mesonbuild/backend/xcodebackend.py]: __[flake8]__
* [W293] blank line contains whitespace

File: mesonbuild/backend/xcodebackend.py [hunk 6]
```diff
@@ -300,6 +334,29 @@ def generate_pbx_file_reference(self):
             self.ofile.write(target_templ % (idval, tname, typestr, path, reftype))
         self.ofile.write('/* End PBXFileReference section */\n')
 
+    def generate_pbx_frameworks_buildphase(self):
+        self.ofile.write('\n/* Begin PBXFrameworksBuildPhase section */\n')
+        self.indent_level += 1
+        self.write_line('%s /* %s */ = {\n' %(self.buildphasemap['Frameworks'], 'Frameworks'))
```
Comment (ignatenkobrain) [inline mesonbuild/backend/xcodebackend.py]: __[flake8]__
* [E225] missing whitespace around operator

File: mesonbuild/backend/xcodebackend.py [hunk 7]
```diff
@@ -300,6 +334,29 @@ def generate_pbx_file_reference(self):
...
+        self.indent_level += 1
+        self.write_line('isa = PBXFrameworksBuildPhase;\n')
+        self.write_line('buildActionMask = %s;\n' %(2147483647))
```
Comment (ignatenkobrain) [inline mesonbuild/backend/xcodebackend.py]: __[flake8]__
* [E225] missing whitespace around operator

File: mesonbuild/backend/xcodebackend.py [hunk 8]
```diff
@@ -300,6 +334,29 @@ def generate_pbx_file_reference(self):
...
+        self.write_line('files = (\n')
+        self.indent_level += 1
+
```
Comment (ignatenkobrain) [inline mesonbuild/backend/xcodebackend.py]: __[flake8]__
* [W293] blank line contains whitespace

File: mesonbuild/backend/xcodebackend.py [hunk 9]
```diff
@@ -300,6 +334,29 @@ def generate_pbx_file_reference(self):
...
+        self.write_line('isa = PBXFrameworksBuildPhase;\n')
+        self.write_line('buildActionMask = %s;\n' %(2147483647))
+        self.write_line('files = (\n')
+        self.indent_level += 1
+        
+        for t in self.build.targets.values():
+            for dep in t.get_external_deps():
+                if isinstance(dep, dependencies.AppleFrameworks):
+                    for f in dep.frameworks:
+                        self.write_line('%s /* %s.framework in Frameworks */,\n' %(self.native_frameworks[f], f))
```
Comment (ignatenkobrain) [inline mesonbuild/backend/xcodebackend.py]: __[flake8]__
* [E225] missing whitespace around operator

File: mesonbuild/backend/xcodebackend.py [hunk 10]
```diff
@@ -300,6 +334,29 @@ def generate_pbx_file_reference(self):
...
+
```
Comment (ignatenkobrain) [inline mesonbuild/backend/xcodebackend.py]: __[flake8]__
* [W293] blank line contains whitespace

File: mesonbuild/backend/xcodebackend.py [hunk 11]
```diff
@@ -348,6 +407,26 @@ def generate_pbx_group(self):
         self.write_line('sourceTree = "<group>";')
         self.indent_level -= 1
         self.write_line('};')
+
```
Comment (ignatenkobrain) [inline mesonbuild/backend/xcodebackend.py]: __[flake8]__
* [W293] blank line contains whitespace

File: mesonbuild/backend/xcodebackend.py [hunk 12]
```diff
@@ -348,6 +407,26 @@ def generate_pbx_group(self):
...
+        self.write_line('isa = PBXGroup;')
+        self.write_line('children = (')
+        # write frameworks
+        self.indent_level += 1
+
+        for t in self.build.targets.values():
+            for dep in t.get_external_deps():
+                if isinstance(dep, dependencies.AppleFrameworks):
+                    for f in dep.frameworks:
+                        self.write_line('%s /* %s.framework */,\n' %(self.native_frameworks_fileref[f], f))
```
Comment (ignatenkobrain) [inline mesonbuild/backend/xcodebackend.py]: __[flake8]__
* [E225] missing whitespace around operator

File: mesonbuild/backend/xcodebackend.py [hunk 13]
```diff
@@ -348,6 +407,26 @@ def generate_pbx_group(self):
...
+
```
Comment (ignatenkobrain) [inline mesonbuild/backend/xcodebackend.py]: __[flake8]__
* [W293] blank line contains whitespace

**Comment by nirbheek** [comment]
Btw, have you run the tests using the xcode backend yet? We have a lot of failures there, and it would be very nice to get all the tests passing and add the xcode backend to the CI.

**Comment by kalmiya** [comment]
I didn't run any automated tests, what kind of failures are reported?and how do I trigger the tests to run?

**Comment by jpakkane** [comment]
`./run_tests.py`

**Comment by kalmiya** [comment]
The xcodebackend.py contains a reference to "meson_test.py" which does not exist. I don't know what the contents of that file should be, so I created a dummy  "/mesonbuild/scripts/meson_test.py" so it continues... and indeed there are quite a lot of errors, so please wait a bit still with merging this pull-request, I'd like to take a closer look at at least a few of those.

**Comment by nirbheek** [comment]
For meson_test.py, you'll want to look at c7ddce163fb8f440c96649a3319ab9f1f62e38be (and the commits around that). The whole thing was reworked into a new `mesontest.py` entry-point that is always used for running tests and is also available to the user.

**Comment by nirbheek** [comment]
If you want to compare how things work with the Ninja backend, you can run `./run_tests.py --backend=ninja`. Everything should pass with that since we run it on the CI.

Btw, you don't have to fix everything that's broken because I can imagine that's a lot of work. It's a good first step if you can get the OS X backend to a point where most of it works and then we can skip the tests that don't with:

```meson
if meson.backend() == 'xcode'
  error('MESON_SKIP_TEST broken with the xcode backend')
endif
```

Seeing that `error()` message will tell `run_project_tests.py` to skip that test.

Then we can add it to the CI and ensure that we don't get regressions, and we can slowly work on fixing the rest. :)

**Comment by kalmiya** [comment]
Ok, thanks for the info, that's helpful 

The unittests are actually quite useful since the code in my PR was only tested for my specific case and now "common testcase 8" is failing due to my changes; It looks like it's going to be a bit more complex (who'd have expected that ^^). 

I already wrote a comment in the code that we'd probably need frameworks per target and then the loop over self.buildphasemap.items() will have to be adjusted accordingly (only add the correct framework-per-target instead of looping). Since I only have 1 target I simply didn't run into any issues.

Anyway, will take a closer look, but probably earliest in the weekend.

File: mesonbuild/backend/xcodebackend.py [hunk 14]
```diff
@@ -300,6 +331,27 @@ def generate_pbx_file_reference(self):
             self.ofile.write(target_templ % (idval, tname, typestr, path, reftype))
         self.ofile.write('/* End PBXFileReference section */\n')
 
+    def generate_pbx_frameworks_buildphase(self):
+        for tname, t in self.build.targets.items():
+            self.ofile.write('\n/* Begin PBXFrameworksBuildPhase section */\n')
+            self.indent_level += 1
+            self.write_line('%s /* %s */ = {\n' %(t.buildphasemap['Frameworks'], 'Frameworks'))
```
Comment (ignatenkobrain) [inline mesonbuild/backend/xcodebackend.py]: __[flake8]__
* [E225] missing whitespace around operator

File: mesonbuild/backend/xcodebackend.py [hunk 15]
```diff
@@ -300,6 +331,27 @@ def generate_pbx_file_reference(self):
...
+            self.indent_level += 1
+            self.write_line('isa = PBXFrameworksBuildPhase;\n')
+            self.write_line('buildActionMask = %s;\n' %(2147483647))
```
Comment (ignatenkobrain) [inline mesonbuild/backend/xcodebackend.py]: __[flake8]__
* [E225] missing whitespace around operator

**Comment by jpakkane** [comment]
Just let us know when you think this is ready to merge. It does not need to be perfect, just better than what we currently have and no obvious regressions.

**Comment by nirbheek** [comment]
Btw, we recently pushed support for running the `./run_unittests.py` with backends other than Ninja, so you might want to rebase and look at that too.

**Comment by kalmiya** [comment]
Ok, to the best of my knowledge it should work. Tested it with my own App and with a modified unittest "8 install" which is linking with a different framework for both the lib and the exe  (also committed - at the very least that test should compile).

**Comment by nirbheek** [comment]
`./run_unittests.py` is unrelated to your last commit. It's a totally different set of tests.

**Comment by nirbheek** [comment]
To be clear, `./run_tests.py` runs `./run_unittests.py` and `./run_project_tests.py`.

**Comment by kalmiya** [comment]
"./run_unittests.py is unrelated to your last commit. It's a totally different set of tests."
I was more answering to jpakkane concern of 'obvious regressions' by showing the testscenario I used -
and since I had it anyway, it made sense to commit it.

Besides that, I just started checking out the failing unit-tests, specifically the first few, common 4 & 6 and made some notes in issue #1597.

**Comment by kalmiya** [comment]
What is the time-schedule to get this merged?

**Comment by jpakkane** [comment]
This message is posted to every outstanding merge request.

We have transitioned from Github wiki to our new documentation system that generates all documentation directly from the Git repo. This means that from now on every merge request must come with corresponding documentation changes.

If your MR requires documentation changes, do the necessary changes to files in `docs/markdown`.

If your MR adds a new feature, add a description to the release notes in `docs/markdown/Release-notes-for-0.41.0.md`.

If this commit is just a bugfix with no functional changes, you don't need to do anything.

Thank you

**Comment by jpakkane** [comment]
This conflicts currently. Once that is done we can merge whenever you feel it is good enough (and there are no obvious regressions).

**Comment by jpakkane** [comment]
It would be great to get this rebased and merged.

**Pull request merged**